### ISSUE TITLE
KokkosKernels pullout broke my configures

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#1099]
Any idea on what I need to add to my scripts to get configure to work.  It is a silent failure, by that I mean it isn't obvious what the problem is, and my CMakeErrors.log looks pretty much the same before and after the commit. 

I added Trilinos_ENABLE_KokkosKernels=ON

Set up for creating a distribution ...
tail of the configure.
`
Finished configuring Trilinos!

-- Configuring incomplete, errors occurred!
See also "/jenkins/slave/workspace/EMPIRE-debug/build_drekar/CMakeFiles/CMakeOutput.log".
See also "/jenkins/slave/workspace/EMPIRE-debug/build_drekar/CMakeFiles/CMakeError.log".
Error in configuration of drekar, trying to continue`

Here is my configure script
[config.txt](https://github.com/trilinos/Trilinos/files/811733/config.txt)
and all the log files. log has been renamed to txt

[configure.txt](https://github.com/trilinos/Trilinos/files/811735/configure.txt)
[CMakeError.txt](https://github.com/trilinos/Trilinos/files/811737/CMakeError.txt)
[CMakeOutput.txt](https://github.com/trilinos/Trilinos/files/811738/CMakeOutput.txt)

**Comment by aprokop** [comment]
@csiefer2 @crtrott

**Comment by rppawlo** [comment]
I'm seeing it too. Here's the error I get at configure:

Processing enabled package: KokkosKernels (Libs)
CMake Error at packages/kokkos-kernels/CMakeLists.txt:46 (TRIBITS_ADD_OPTION_AND_DEFINE):
  TRIBITS_ADD_OPTION_AND_DEFINE Macro invoked with incorrect arguments for
  macro named: TRIBITS_ADD_OPTION_AND_DEFINE

**Comment by mhoemmen** [comment]
Hey @bathmatt and @rppawlo, does any package not part of Trilinos' public distribution, that you happen to use, depend on KokkosKernels directly (not just indirectly through Tpetra)?  This could explain the issue.  @crtrott did test Trilinos but he may not have tested, say, Panzer.

**Comment by rppawlo** [comment]
I would have expected Intrepid2 to depend on it but it doesn't look like it has that dependency yet. Otherwise, I don't think so.

**Comment by bathmatt** [comment]
-bash-4.1$ find . -name Dependencies.cmake -exec grep -il KokkosKernel {} \;
./packages/ifpack2/cmake/Dependencies.cmake
./packages/stokhos/cmake/Dependencies.cmake
./packages/tpetra/core/cmake/Dependencies.cmake
./packages/trilinoscouplings/cmake/Dependencies.cmake

**Comment by bathmatt** [comment]
Hereis the matching grep

SET(LIB_REQUIRED_DEP_PACKAGES Belos Teuchos Tpetra KokkosKernels)
SET(LIB_OPTIONAL_DEP_PACKAGES EpetraExt Ifpack ML TriKota Anasazi Sacado NOX Isorropia KokkosKernels TeuchosKokkosComm KokkosAlgorithms KokkosContainers Tpetra Ifpack2 MueLu Belos Amesos2 Thyra)
  LIB_REQUIRED_PACKAGES Teuchos TpetraClassic KokkosCore KokkosContainers KokkosAlgorithms TeuchosKokkosCompat TeuchosKokkosComm KokkosKernels
SET(TEST_OPTIONAL_DEP_PACKAGES Amesos Isorropia Epetra EpetraExt Ifpack Intrepid Intrepid2 Pamgen AztecOO ML Zoltan STKClassic Teko KokkosKernels Tpetra MueLu KokkosCore TeuchosKokkosCompat KokkosContainers TeuchosKokkosComm Stokhos TriKota Sacado)

**Comment by mhoemmen** [comment]
@bathmatt We _want_ to see KokkosKernels.  We don't want to see TpetraKernels.
```
$ for file in $(find . -name Dependencies.cmake); do grep "TpetraKernels" $file; done
$
$ for file in $(find . -name Dependencies.cmake); do grep "KokkosKernels" $file; done
SET(LIB_REQUIRED_DEP_PACKAGES Belos Teuchos Tpetra KokkosKernels)
SET(LIB_OPTIONAL_DEP_PACKAGES EpetraExt Ifpack ML TriKota Anasazi Sacado NOX Isorropia KokkosKernels TeuchosKokkosComm KokkosAlgorithms KokkosContainers Tpetra Ifpack2 MueLu Belos Amesos2 Thyra)
  LIB_REQUIRED_PACKAGES Teuchos TpetraClassic KokkosCore KokkosContainers KokkosAlgorithms TeuchosKokkosCompat TeuchosKokkosComm KokkosKernels
SET(TEST_OPTIONAL_DEP_PACKAGES Amesos Isorropia Epetra EpetraExt Ifpack Intrepid Intrepid2 Pamgen AztecOO ML Zoltan STKClassic Teko KokkosKernels Tpetra MueLu KokkosCore TeuchosKokkosCompat KokkosContainers TeuchosKokkosComm Stokhos TriKota Sacado)
```

**Comment by bathmatt** [comment]
@mhoemmen I was answering your question here.

"Does any package not part of Trilinos' public distribution, that you happen to use, depend on KokkosKernels directly "

tpetra and ifpack2 both do.

**Comment by mhoemmen** [comment]
Hm, this is interesting.  In my Clang 3.5 build:
```
Processing enabled package: KokkosKernels (Libs)
CMake Error at packages/kokkos-kernels/CMakeLists.txt:46 (TRIBITS_ADD_OPTION_AND_DEFINE):
  TRIBITS_ADD_OPTION_AND_DEFINE Macro invoked with incorrect arguments for
  macro named: TRIBITS_ADD_OPTION_AND_DEFINE
```

**Comment by bathmatt** [comment]
@mhoemmen 
yes, @rppawlo reported the same thing a bit ago.  So, I don't think it worked for anyone.

his comment
"I'm seeing it too. Here's the error I get at configure:

Processing enabled package: KokkosKernels (Libs)
CMake Error at packages/kokkos-kernels/CMakeLists.txt:46 (TRIBITS_ADD_OPTION_AND_DEFINE):
TRIBITS_ADD_OPTION_AND_DEFINE Macro invoked with incorrect arguments for
macro named: TRIBITS_ADD_OPTION_AND_DEFINE"

**Comment by mhoemmen** [comment]
> @mhoemmen I was answering your question here.
> 
> "Does any package not part of Trilinos' public distribution, that you happen to use, depend on KokkosKernels directly "
> 
> tpetra and ifpack2 both do.

Hi @bathmatt -- sorry, I should have emphasized "not part of Trilinos' public distribution."  Tpetra and Ifpack2 are part of Trilinos' public distribution.  You folks use packages that are not Trilinos packages, but integrate with Trilinos through TriBITS.

**Comment by mhoemmen** [comment]
OK, let me take a look.  Maybe this is easy.

**Comment by mhoemmen** [comment]
I think it's this _ENABLE_EXECUTION_SPACE_ variable.  That's not defined, I don't think, which confuses CMake.  Also, the ASSERT_DEFINED on line 45, right above the use of one of those variables, refers to `Kokkos_ENABLE_CUDA`, not `${PACKAGE_NAME}_ENABLE_EXECUTION_SPACE_CUDA`.

**Comment by mhoemmen** [comment]
I get it.  @crtrott wants to be able to control the execution spaces for which KokkosKernels prebuilds kernels, separately from the execution spaces that Kokkos enables.  I think I can fix this by setting defaults for those variables if they do not exist.

**Comment by mhoemmen** [comment]
I'm testing the fix now.

**Comment by mhoemmen** [comment]
Ok, I think I got it.  Give me a few minutes to push.  I'll push the fix to Trilinos develop first, then prepare a PR for KokkosKernels develop.

**Comment by mhoemmen** [comment]
I just pushed the fix to Trilinos develop.

**Comment by rppawlo** [comment]
Thanks Mark!

**Comment by bathmatt** [comment]
thanks!!

**Comment by bathmatt** [comment]
@mhoemmen did you get it to compile?  I'm getting tons of errors now and nothing is compiling?

**Comment by rppawlo** [comment]
Same for me. Must be the defaults are incorrect. I get tons of errors like the one below. I don't have Cuda enabled. It is an OpenMP only build.

[ 47%] Building CXX object packages/epetra/src/CMakeFiles/epetra.dir/Epetra_SerialSymDenseMatrix.cpp.o
/ascldap/users/rppawlo/Trilinos/packages/kokkos-kernels/src/impl/Kokkos_Sparse_impl_spmv_decl.hpp:258:53: error: ‘Cuda’ is not a member of ‘Kokkos’
   KOKKOSSPARSE_IMPL_SPMV_DEFAULTS_DECL( SCALAR, LO, Kokkos::Cuda, Kokkos::CudaUVMSpace )
                                                     ^
/ascldap/users/rppawlo/Trilinos/packages/kokkos-kernels/src/impl/Kokkos_Sparse_impl_spmv_decl.hpp:188:85: note: in definition of macro KOKKOSSPARSE_IMPL_SPMV_DEFAULTS_DECL’
             Kokkos::StaticCrsGraph<ORDINAL_TYPE, Kokkos::LayoutLeft, Kokkos::Device<EXEC_SPACE_TYPE, MEM_SPACE_TYPE> >::size_type, \
                                                                                     ^
/ascldap/users/rppawlo/BUILD/packages/kokkos-kernels/src/KokkosKernels_ETIHelperMacros.h:15:2: note: in expansion of macro ‘KOKKOSSPARSE_IMPL_SPMV_DEFAULTS_DECL_CUDA’
  INSTMACRO( double , int )
  ^
/ascldap/users/rppawlo/Trilinos/packages/kokkos-kernels/src/impl/Kokkos_Sparse_impl_spmv_decl.hpp:260:1: note: in expansion of macro ‘KOKKOSKERNELS_INSTANTIATE_SL’
 KOKKOSKERNELS_INSTANTIATE_SL( KOKKOSSPARSE_IMPL_SPMV_DEFAULTS_DECL_CUDA )
 ^
/ascldap/users/rppawlo/Trilinos/packages/kokkos-kernels/src/impl/Kokkos_Sparse_impl_spmv_decl.hpp:400:55: error: template argument 1 is invalid
         Kokkos::Device<EXEC_SPACE_TYPE, MEM_SPACE_TYPE>, \
                                                       ^

**Comment by csiefer2** [comment]
Yeah, same here.

**Comment by mhoemmen** [comment]
Ok, that's weird.  Could you send me your CMakeCache.txt files?

**Comment by crtrott** [comment]
Hey guys, my feeling is this is two different errors. Matts error (looking at the actual error output seems to be an issue with BLAS while Rogers was something with cmake.

**Comment by crtrott** [comment]
Ok I got the build error fixed. The attempted fix by Mark was basically always enabling all backends inside of KokkosKernels. There need to be some if clauses.

**Comment by crtrott** [comment]
I still don't know what the initial error is though the output doesn't say anything ...

**Comment by crtrott** [comment]
Btw on which machines was this?

**Comment by crtrott** [comment]
Ok this should have added a proper fix.

**Comment by rppawlo** [comment]
yes - that did it! Thanks!

**Comment by mhoemmen** [comment]
I don't understand why my previous commit didn't fix things, but I'm glad it works now!  Thanks @crtrott ! :-D

**Comment by bathmatt** [comment]
@crtrott thanks for the fix, it works for me now.

The configure was broken on three+ platforms.  my machine, jenkins RHEL6 (which is like my machine) shiller and ellis.  Not sure how it worked anywhere before Mark's fix.

Closing

**Comment by csiefer2** [comment]
Can build now, thanks.

**Comment by jjellio** [comment]
I am on develop, and I cannot use an installed Trilinos build. 
```
In file included from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Kokkos_Sparse_impl_spmv.hpp(48),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Kokkos_Sparse.hpp(63),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Tpetra_CrsMatrix_decl.hpp(63),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Tpetra_CrsMatrix.hpp(1),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Xpetra_TpetraCrsMatrix.hpp(57),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Xpetra_CrsMatrixFactory.hpp(54),
                 from /home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Xpetra_Matrix.hpp(60),
                 from /home/jjellio/src/trilinos-perf/SolverDriver/SolverDriverDetails_decl.hpp(77),
                 from /home/jjellio/src/trilinos-perf/SolverDriver/SolverDriver.cpp(77):
/home/jjellio/install/trilinos-12.11/intel-17.0.1_mpt-7.4.2_hsw-openmp/include/Kokkos_Sparse_impl_spmv_def.hpp(50): catastrophic error: cannot open source file "impl/Kokkos_Sparse_impl_spmv_omp.hpp"
  #include "impl/Kokkos_Sparse_impl_spmv_omp.hpp"
                                                 ^
```

**Comment by ibaned** [comment]
@jjellio  I ran into the same thing, and submitted a fix as kokkos/kokkos-kernels#8

**Comment by jjellio** [comment]
Nice, I just copied the files manually... The machine I am on is going down for a week and I needed things going fast.
